CompareRaftMembers/DiffStreamOplogFilter.js

```diff
@@ -0,0 +1,240 @@
+        this.maxBufferedEntries = userMaxBufferedEntries || DEFAULT_MAX_BUFFERED_ENTRIES;
+        const retryDelayMs = userRetryDelayMs || DEFAULT_RETRY_DELAY_MS;
+        const maxRetryDelayMs = userMaxRetryDelayMs || DEFAULT_MAX_RETRY_DELAY_MS;
+        this.retryParams = {
+            times: 20,
+            interval: retryCount => Math.min(
+                // the first retry comes as "retryCount=2", hence substract 2
+                retryDelayMs * (2 ** (retryCount - 2)),
+                maxRetryDelayMs,
+            ),
+        };
+        this.bucketNameToRaftSessionId = {};
+        this._initRaftOplogStreams(params);
+
+        this.nBufferedEntries = 0;
+        this.pendingTransformCallback = null;
+        this.pendingFlushCallback = null;
+    }
+
+    _initRaftOplogStreams(params) {
+        const {
+            bucketdHost, bucketdPort,
+            excludeFromCseqs,
+        } = params;
+
+        this.raftSessionStates = {};
+        for (const rsId of Object.keys(excludeFromCseqs)) {
+            const raftSessionState = {};
+            this.raftSessionStates[rsId] = raftSessionState;
+            const oplogStream = new RaftOplogStream({
+                bucketdHost,
+                bucketdPort,
+                raftSessionId: rsId,
+                startSeq: excludeFromCseqs[rsId] + 1,
+            });
+            this._setupOplogStream(raftSessionState, oplogStream);
+            raftSessionState.oplogStream = oplogStream;
+            raftSessionState.oplogKeys = new Set();
+            raftSessionState.inputBuffer = [];
+        }
+    }
+
+    _setupOplogStream(raftSessionState, oplogStream) {
+        // no 'end' event handling is required since the oplog streams never end
+        oplogStream
+            .on('data', data => this._onOplogEvent(raftSessionState, data))
+            .on('error', err => {
+                this.emit('error', err);
+            });
+    }
+
+    _onOplogEvent(raftSessionState, data) {
+        const { entry } = data;
+        if (entry === null) {
+            // a null entry means we reached the end of
+            // the raft oplog, this is the point where
+            // it becomes safe to process the raft
+            // session's current input buffer
+            this._processInputBuffer(raftSessionState);
+        } else if (entry.method === 'BATCH') {
+            // only consider events from BATCH records
+            const fullKey = `${entry.bucket}/${entry.key}`;
+            raftSessionState.oplogKeys.add(fullKey);
+        }
+    }
+
+    _transform(diffEntry, encoding, callback) {
+        const bucketName = this._extractBucketName(diffEntry);
+        async.waterfall([
+            next => {
+                const rsId = this.bucketNameToRaftSessionId[bucketName];
+                if (rsId !== undefined) {
+                    return next(null, rsId);
+                }
+                return this._fetchBucketRaftSessionId(bucketName, next);
```

Comment:
I would just move lines the behavior of 97-100 inside of _fetchBucketRaftSessionId